ed_sensor_integration/src/laser/updater.cpp

```diff
@@ -0,0 +1,655 @@
+#include <ed/laser/updater.h>
+
+#include <iostream>
+
+#include <geolib/Shape.h>
+
+#include <ed/world_model.h>
+#include <ed/entity.h>
+#include <ed/update_request.h>
+
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include "ed/convex_hull_calc.h"
+
+#include <ed/io/json_writer.h>
+
+#include "ed_sensor_integration/association_matrix.h"
+
+#include <tue/profiling/timer.h>
+
+/**
+ * Calculate an error based on the quality of a fit.
+ *
+ * @param e entity being fitted
+ * @param lrf laser range finder
+ * @param rel_pose candidate pose of the entity relative to the laser range finder
+ * @param sensor_ranges distances measured by the lrf
+ * @param model_ranges predicted measurement distances using the model sans e
+ * @param[out] num_model_points number of points used for
+ * @return double error
+ */
+double getFittingError(const ed::Entity& e, const geo::LaserRangeFinder& lrf, const geo::Pose3D& rel_pose,
+                       const std::vector<double>& sensor_ranges, const std::vector<double>& model_ranges,
+                       int& num_model_points)
+{
+    std::vector<double> test_model_ranges = model_ranges;
+
+    // Render the entity with the given relative pose
+    geo::LaserRangeFinder::RenderOptions opt;
+    opt.setMesh(e.shape()->getMesh(), rel_pose);
+
+    geo::LaserRangeFinder::RenderResult res(test_model_ranges);
+    lrf.render(opt, res);
+
+    int n = 0;
+    num_model_points = 0;
+    double total_error = 0;
+    for(unsigned int i = 0; i < test_model_ranges.size(); ++i)
+    {
+        double ds = sensor_ranges[i];
+        double dm = test_model_ranges[i];
+
+        if (ds <= 0) // no sensor data
+            continue;
+
+        ++n;
+
+        if (dm <= 0) // no raytrace collision in model
+        {
+            total_error += 0.1;
+            continue;
+        }
+
+        double diff = std::abs(ds - dm);
+        if (diff < 0.1)
+            total_error += diff;
+        else
+        {
+            if (ds > dm)
+                total_error += 1;
+            else
+                total_error += 0.1;
+        }
+
+        ++num_model_points;
+    }
+
+    return total_error / (n+1); // to be sure to never divide by zero.
```

Comment:
```suggestion
    return total_error / (n+1); // To be sure to never divide by zero.
```
Is this behaviour desired over `max(n, 1)`?

Reply:
It makes no difference for the algorithm and I believe n+1 is faster than max